### ISSUE TITLE
Random Photo Colors and Sizes

### ISSUE THREAD

**Issue by saulshanabrook** [rbarrois/factory_boy#260]
I see that I can create a photo with a certain color and size [like this](https://factoryboy.readthedocs.org/en/latest/orms.html?highlight=django#factory.django.ImageField):

```python
class MyFactory(factory.django.DjangoModelFactory):
    class Meta:
        model = models.MyModel

    the_image = factory.django.ImageField(color='blue')
```

Is it possible to randomize that color and also the dimensions?

More generally, this means using one declaration as the input to another declaration in generation.

**Comment by rbarrois** [comment]
Indeed, using one declaration as an input to another one would be the best way to solve this.

That's also a very nice feature I've had in mind for quite some time :)

Would you be interested in building this? I don't have much time available to work on this in the near future, but I'd be more than happy to guide you through the codebase if you're interested !

**Comment by saulshanabrook** [comment]
Yeah I think I would be. Do you have any thoughts on how to design this API? I will lay out what I have been thinking so far. 

## Use Cases

Like I said above, I want to be able to generate an `ImageField` with random colors and sizes. It seems like the [`FuzzyChoice`](https://factoryboy.readthedocs.org/en/latest/fuzzy.html#fuzzychoice) and [`FuzzyInteger`](https://factoryboy.readthedocs.org/en/latest/fuzzy.html#fuzzyinteger) declerations could be used to achieve this, but currently there isn't a straightforward way to achieve this.

Also, on one occasion I wanted to use the `FuzzyDate`, but make the `start_date` equal to another field, so that both fields are randomized, but one has to be after the other.

We can consider this (minimal) Django model as an example:

```python
class Exhibition(models.Model):
    start_date = models.DateField()
    end_date = models.DateField()
    image = models.ImageField()
``` 

## Possible APIs

### Option 1

One option is to allow `LazyAttribute` (and possibly `Sequence` and `LazyAttributeSequence` to be comprehensive) to return declarations instead of absolute values.

```python
class ExhibitionFactory(factory.DjangoModelFactory):
    class Meta:
        model = Exhibition
        exclude = ('image_color', 'image_height', 'image_width')

    start_date = factory.fuzzy.FuzzyDate()
    end_date = factory.LazyAttribute(lambda obj: factory.fuzzy.FuzzyDate(start_date=obj.start_date))

    image_color = factory.fuzzy.FuzzyChoice(['blue', 'yellow', 'green', 'orange'])
    image_height = factory.fuzzy.FuzzyInteger(10, 1000)
    image_width = factory.fuzzy.FuzzyInteger(10, 1000)

    image = factory.LazyAttribute(
        lambda obj: factory.django.ImageField(color=obj.image_color, height=obj.image_height, width=obj.image_width)
    )
```

This looks nice for the dates, but it is rather verbose for the image. Also, it would be hard to separate out this randomized `ImageField` to make it reusable across factories, because it requires adding other fields and making sure they are excluded from the build.

### Option 2

Alternatively, we could allow the arguments passed to declarations to be declarations themselves.

```python
class ExhibitionFactory(factory.DjangoModelFactory):
    class Meta:
        model = Exhibition

    start_date = factory.fuzzy.FuzzyDate()
    end_date = factory.FuzzyDate(start_date=factory.SelfAttribute('start_date'))

    image = factory.django.ImageField(
        color=factory.fuzzy.FuzzyChoice(['blue', 'yellow', 'green', 'orange']),
        height=factory.fuzzy.FuzzyInteger(10, 1000),
        width=factory.fuzzy.FuzzyInteger(10, 1000),
    )
```

This is more succinct and this randomized `ImageField` could easily be reused.

**Comment by rbarrois** [comment]
Awesome!

I like the second option better: it has less clutter, and feels consistent with ``factory.SubFactory``.
Also, this means that there is already some support for these features, namely in ``factory.declarations.ParameteredAttribute`` (the base class of ``factory.SubFactory``).

Now, I also see a design decision to take: what is the "current" context in such a declaration?

In other words, if I write:

```python
class ExhibitionFactory(factory.django.DjangoModelFactory):
    class Meta:
        model = Exhibition

    start_date = factory.fuzzy.FuzzyDate()
    end_date = factory.fuzzy.FuzzyDate(
        start_date=factory.SelfAttribute('start_date'),
        end_date=factory.LazyAttribute(lambda d: d.start_date + datetime.timedelta(days=10)),
    )

    image = factory.django.ImageField(
        width=factory.fuzzy.FuzzyInteger(10, 1000),
        height=factory.SelfAttribute('width'),
    )
```

Does the ``d.start_date`` refer to the ``start_date`` of the ``FuzzyDate`` or of the ``ExhibitionFactory``?
In a similar manner, should ``image.height`` copy ·``image.width`` or ·``exhibition.width``?

**Comment by saulshanabrook** [comment]
Interesting question. My default inclination is to assume that `ExhibitionFactory` is the "current". However, I think if it was implemented this way, then having an image with a random height, and the same width, would be hard, because you would have to say something like `height=factory.SelfAttribute('image.width')` which doesn't seem like would be possible, because the image can't be created till the height is specified, but the height depends on knowing the image width. It might be possible, because we could calculate the width before creating the image, but it seems more complicated.

So I think it seems simpler to have it refer to the parent.

This also would make it similar to [how it works with a `SubFactory`](https://factoryboy.readthedocs.org/en/latest/reference.html#parents).

Looking at the `SubFactory`, another option is to create a new declaration called something like `DynamicDeclaration` to use like this:

```
class ExhibitionFactory(factory.django.DjangoModelFactory):
    class Meta:
        model = Exhibition

    start_date = factory.fuzzy.FuzzyDate()
    end_date = factory.DynamicDeclaration(
        factory.fuzzy.FuzzyDate,
        start_date=factory.SelfAttribute('start_date'),
        end_date=factory.LazyAttribute(lambda end_date: end_date.parent.start_date + datetime.timedelta(days=10)),
    )

    image = factory.DynamicDeclaration(
        factory.django.ImageField,
        width=factory.fuzzy.FuzzyInteger(10, 1000),
        height=factory.SelfAttribute('width'),
    )
```

**Comment by rbarrois** [comment]
Well, since all those declarations are already explicit "factoryboy-related" through their import path, I find the ``SubFactory`` style simpler and more intuitive.

Using a ``DynamicDeclaration`` would be weird when mixing lazy values with explicit ones, eg:
```python
image = factory.DynamicDeclaration(
    factory.django.ImageField,
    color='red',
    width=factory.fuzzy.FuzzyInteger(10, 1000),
    height=factory.SelfAttribute('width'),
)
```

So I'd rather go for the direct way; i.e:

```python
class ExhibitionFactory(factory.django.DjangoModelFactory):
    class Meta:
        model = Exhibition

    start_date = factory.fuzzy.FuzzyDate()
    end_date = factory.fuzzy.FuzzyDate(
        start_date=factory.SelfAttribute('..start_date'),
        end_date=factory.LazyAttribute(lambda exhibition_end: exhibition_end.start_date + datetime.timedelta(days=10)),
    )

    image = factory.django.ImageField(
        width=factory.fuzzy.FuzzyInteger(10, 1000),
        height=factory.SelfAttribute('width'),
    )

**Comment by rbarrois** [comment]
Hey @saulshanabrook !

Coming back to this after far too long, I was wondering if you're still interested in working on that topic?

Let me know if there is anything I can do to help!

**Comment by saulshanabrook** [comment]
@rbarrois I have moved on to other things at the moment, so working on this feature isn't a priority.

**Comment by rbarrois** [comment]
@saulshanabrook OK, sorry to hear that :-(

And thanks for the help in defining the API for that new feature :)

**Comment by saulshanabrook** [comment]
Yeah me too, it seems like a fun feature to work on.

**Comment by hvdklauw** [comment]
I would really like this feature too, I'll see if I can free up some time for this.

**Comment by hvdklauw** [comment]
Hmm after looking it some more; doing it the direct way would mean the OrderedDeclaration would need to do the preprocessing of parameters, but it has nothing to do so.

Using a DynamicDeclaration with a simple isinstance on OrderedDeclaration would be a lot simpler, if it is an OrderedDeclaration evaluate it and update the params before passing it to the passed attribute which you then evaluate in turn.

Something like this (pseudocode):

```python
class DynamicDeclaration(OrderedDeclarations):


    def __init__(self, field, **kwargs):
        super(DynamicDeclaration, self).__init__()
        self.field = field
        self.defaults = kwargs

    def evaluate(self, sequence, obj, create, extra=None, containers=()):
        params = {}
        for key in self.defaults:
            value = self.defaults[key]
            if isinstance(value, OrderedDeclarations):
                value = value.evaluate(sequence, obj, create, extra=extra, containers=containers + (obj,))
            params[key] = value
        return self.field(**params).evaluate(sequence, obj, create, extra=extra, containers=containers)
```

**Comment by rbarrois** [comment]
Hey @hvdklauw , welcome to the project :)

Your approach sounds good to me, and actually it looks like the current implementation of a factory, in ``factory.containers.AttributeBuilder.build()``.

It should be possible to refactor that part of the code to extract the "building a dict of parameters" part, don't you think?

**Comment by hvdklauw** [comment]
I see a lof of similarities, but I need some time to process how they might apply.

**Comment by jeffwidman** [comment]
Quick FYI:
Per the discussion in #271, many of the built-in fuzzers for factory_boy are going to be deprecated/eventually removed in favor of their Faker equivalents. 

This feature still seems like it'd be useful/interesting, just something to keep in mind.

**Comment by rblenkinsopp** [comment]
Is there any update on the current preferred way to do this? I'm looking to populate a factory.django.FileField with a random file name and data content, and was attempting something similar to:

`class FileFactory(factory.django.DjangoModelFactory):
    class Meta:
        model = 'test.File'
        exclude = ('file_name', 'file_size', 'file_data')

    file_name = factory.Faker('file_name')
    file_size = factory.Faker('random_int', min=1, max=1048576)
    file_data = factory.LazyAttribute(lambda o: factory.Faker('binary', length=o.file_size))
    file = factory.LazyAttribute(lambda o: factory.django.FileField(filename=o.file_name, data=o.file_data))`

But this is currently failing with a rather obscure error message:

`AttributeError: 'FileField' object has no attribute '_committed'`

**Comment by x-yuri** [comment]
I've found the same can be done by overriding `generate` method:

```py
class RandomImageField(factory.django.ImageField):
    def generate(self, step, params):
        for n in ['width', 'height']:
            if not n in params:
                continue
            v = params[n]
            if callable(v):
                v = v()
            params[n] = v
        return super(RandomImageField, self).generate(step, params)

class PhotoFactory(factory.django.DjangoModelFactory):
    class Meta:
        model = Photo
    file = RandomImageField(
        width=faker().image_width,
        height=faker().image_height)

def faker():
    return factory.Faker._get_faker()

class MyProvider(BaseProvider):
    def image_width(self):
        return self.random_int(300, 400)

    def image_height(self):
        return self.random_int(400, 500)

factory.Faker.add_provider(MyProvider)

```

Seeing how `generate` doesn't start with an underscore make it feel like a more legitimate/safe way. But if you want to persist the generated values in the model (it has corresponding fields), you can replace:

```py
            params[n] = v
```

with:

```py
            step.attributes[n] = params[n] = v
```
